File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root: return []
        ans = []
        
        def helper(node,level):
            if not node: return
            
            if len(ans) == level: ans.append(node.val)
            
            if node.right:
                helper(node.right, level+1)
            if node.left:
                helper(node.left, level+1)
                
        helper(root, 0)
        return ans
#         bfs. time: o(n), space: o(n)
#         if not root: return []
#         que = [[root, 0]]
#         ans = []
        
#         while que:
#             [curNode, level] = que.pop(0)
           
#             if curNode.right:
#                 que.append([curNode.right, level+1])
#             if curNode.left:
#                 que.append([curNode.left, level+1])
                
#             if len(ans) == level:
#                 ans.append(curNode.val)
                
#         return ans

File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function(root) {
    
    //bfs
//     if(!root) return [];
//     let queue = [[root, 0]];
//     let ans = [];
    
//     while(queue.length){
//         let [curNode, level] = queue.shift();
//         if(ans.length === level) ans.push(curNode.val);
        
//         if(curNode.right) queue.push([curNode.right, level+1])
//         if(curNode.left) queue.push([curNode.left, level+1]);
        

//     }
//     return ans
       
    //dfs
    if(!root) return [];
    let ans = [];
    helper(root, 0);
    return ans;
    
    function helper(node, level){
        if(!node) return;
        
        if(ans.length === level) ans.push(node.val)
                helper(node.right, level+1)

        helper(node.left, level+1)

    } 
        
    
};